File: 2022/12.04.22/puzzle2.nim
import std/strutils

let input = splitLines(readFile("input.txt"))

proc contain(line: string): bool =
    var nums =  split(line, {'-',','})
    var a = parseInt(nums[0]) <= parseInt(nums[2]) and parseInt(nums[1]) >= parseInt(nums[2])
    var b = parseInt(nums[2]) <= parseInt(nums[0]) and parseInt(nums[3]) >= parseInt(nums[0])
    return a or b

var ct: int
for line in input:
    if contain(line):
        ct+=1

echo ct

File: 2023/12.04.2023/puzzle2.py

import numpy as np

with open("input.txt", "r") as f:
    lines = f.read().splitlines()
    
def calc_copies(line):
    line = line.split(": ")[1]
    nums = line.split(" | ")
    winning_nums = [int(num) for num in nums[0].split()]
    my_nums = [int(num) for num in nums[1].split()]
    ct = 0
    for num in winning_nums:
        if num in my_nums:
            ct+=1
    return ct

cards = np.ones(len(lines), int)

for idx, line in enumerate(lines):
    copies = calc_copies(line)
    cards[idx+1:idx+copies+1] += cards[idx]

print(cards.sum())